datastore/src/main/java/org/jboss/pnc/datastore/repositories/ProductMilestoneRepositoryImpl.java

```diff
@@ -195,4 +196,54 @@ public List<Tuple> getMilestonesSharingDeliveredArtifacts(Integer milestoneId) {
 
         return entityManager.createQuery(query).getResultList();
     }
+
+    @Override
+    public List<Artifact> getDeliveredArtifactsSharedInMilestones(
+            PageInfo pageInfo,
+            Integer milestone1Id,
+            Integer milestone2Id) {
+        TypedQuery<Artifact> typedQuery = getDeliveredArtifactsSharedInMilestonesQuery(milestone1Id, milestone2Id);
+        if (pageInfo != null) {
+            typedQuery.setFirstResult(pageInfo.getElementOffset());
+            typedQuery.setMaxResults(pageInfo.getPageSize());
+        }
+
+        return typedQuery.getResultList();
+    }
+
+    @Override
+    public int countDeliveredArtifactsSharedInMilestones(Integer milestone1Id, Integer milestone2Id) {
+        TypedQuery<Artifact> typedQuery = getDeliveredArtifactsSharedInMilestonesQuery(milestone1Id, milestone2Id);
+
+        return typedQuery.getResultList().size();
+    }
+
+    private TypedQuery<Artifact> getDeliveredArtifactsSharedInMilestonesQuery(
```

Comment:
Please, redo this query through queryWithPredicates + predicates. You're returning a regular db model, there's no need for custom queries. You'll get both count() and query with that at the same time; + support for filters/sorting. For reference, joins+group+having with predicates were done f.e. here [1].

[1] https://github.com/project-ncl/pnc/blob/1721880b2fef262c83e3fccd2c285432cd30a2b0/facade/src/main/java/org/jboss/pnc/facade/providers/BuildProviderImpl.java#L247

Custom queries are suited for corner cases when you're returning Tuple or when the query is incompatible with predicates.

Reply:
true